src/daft-dsl/src/optimization.rs

```diff
@@ -58,144 +43,33 @@ pub fn requires_computation(e: &Expr) -> bool {
-                replace_columns_with_expressions(child, replace_map).into(),
-                *mode,
-            )),
-            AggExpr::Sum(child) => Expr::Agg(AggExpr::Sum(
-                replace_columns_with_expressions(child, replace_map).into(),
-            )),
-            AggExpr::Mean(child) => Expr::Agg(AggExpr::Mean(
-                replace_columns_with_expressions(child, replace_map).into(),
-            )),
-            AggExpr::Min(child) => Expr::Agg(AggExpr::Min(
-                replace_columns_with_expressions(child, replace_map).into(),
-            )),
-            AggExpr::Max(child) => Expr::Agg(AggExpr::Max(
-                replace_columns_with_expressions(child, replace_map).into(),
-            )),
-            AggExpr::List(child) => Expr::Agg(AggExpr::List(
-                replace_columns_with_expressions(child, replace_map).into(),
-            )),
-            AggExpr::Concat(child) => Expr::Agg(AggExpr::List(
-                replace_columns_with_expressions(child, replace_map).into(),
-            )),
-        },
-        Expr::BinaryOp { left, right, op } => Expr::BinaryOp {
-            op: *op,
-            left: replace_columns_with_expressions(left, replace_map).into(),
-            right: replace_columns_with_expressions(right, replace_map).into(),
-        },
-        Expr::Cast(child, name) => Expr::Cast(
-            replace_columns_with_expressions(child, replace_map).into(),
-            (*name).clone(),
-        ),
-        Expr::Function { inputs, func } => Expr::Function {
-            func: func.clone(),
-            inputs: inputs
-                .iter()
-                .map(|e| replace_columns_with_expressions(e, replace_map))
-                .collect(),
-        },
-        Expr::Not(child) => replace_columns_with_expressions(child, replace_map),
-        Expr::IsNull(child) => replace_columns_with_expressions(child, replace_map),
-        Expr::IfElse {
-            if_true,
-            if_false,
-            predicate,
-        } => Expr::IfElse {
-            if_true: replace_columns_with_expressions(if_true, replace_map).into(),
-            if_false: replace_columns_with_expressions(if_false, replace_map).into(),
-            predicate: replace_columns_with_expressions(predicate, replace_map).into(),
-        },
+impl<'a> TreeNodeRewriter for ColumnExpressionRewriter<'a> {
+    type N = Expr;
+    fn pre_visit(&mut self, node: &Self::N) -> DaftResult<RewriteRecursion> {
+        if let Expr::Column(name) = node && self.mapping.contains_key(name.as_ref()) {
+            Ok(RewriteRecursion::Continue)
+        } else {
+            Ok(RewriteRecursion::Skip)
+        }
+    }
+    fn mutate(&mut self, node: Self::N) -> DaftResult<Self::N> {
+        if let Expr::Column(ref name) = node && let Some(tgt) = self.mapping.get(name.as_ref()){
+            Ok(tgt.clone())
+        } else {
+            Ok(node)
+        }
```

Comment:
Using a `TreeNodeRewriter` implementation seems more prudent here, although we could also use `expr.transform()` (aka `expr.transform_up()`) with a closure that closes over the column replacement map to do the post-order traversal. E.g.

```rust
pub fn replace_columns_with_expressions(expr: &Expr, replace_map: &HashMap<String, Expr>) -> Expr {
    expr.transform(&|e| {
        if let Expr::Column(ref name) = e && let Some(tgt) = replace_map.get(name.as_ref()) {
            Transformed::Yes(tgt.clone())
        } else {
            Transformed::No(e)
        }
    })
```

This one is more up to personal preference, although I personally prefer the `expr.transform()` since it's more concise and readable to me.

Reply:
I had that but reverted it. The thing I was confused about here is that I didn't see how `Transformed::No` allows the top level Expr to remained unchanged.